### ISSUE TITLE
Cannot assemble correct binary for Polaris 10

### ISSUE THREAD

**Issue by wolf9466** [CLRX/CLRX-mirror#6]
Since I'm unable to use the LLVM toolchain, this is the only way I can properly work on kernels - it'd be nice to have Polaris 10 support.

**Comment by matszpk** [comment]
Can you give some failed examples (source code or excerpts)?

**Comment by wolf9466** [comment]
Anything I disassemble with CLRX - the only example I have right now is long, but I could make a simple anything and compile it using the AMDGPU OCL runtime, and then disassemble, reassemble, and try to load it - says failed to load HSA code object.

**Comment by matszpk** [comment]
Thank you. Can you give me a driver kind and driver version which you are using? I didn't check all devices types given by latest Windows/Linux drivers, so problem can be occurred.

**Comment by wolf9466** [comment]
16.30 AMDGPU drivers - Linux, but since the binaries are cross compatible, I would expect the same from Windows. Polaris 10 (Ellesmere) was my target.

**Comment by matszpk** [comment]
Thank you for answer. I got this same error while trying to run simple example (GenericPtr and VectorAdd) when OpenCL 2.0 code enabled. clinfo got info that my device (Bonaire) doesn't support OpenCL 2.0 (only OpenCL 1.2 C compiler supported). Check whether AMD GPU PRO supports OpenCL 2.0 for your device. I will be trying to other configs to know how these drivers behaves with CLRX.

**Comment by wolf9466** [comment]
It works with -cl-std=CL2.0, but I can try different options. From what I see in your disassembler's output, it's using 2.0 automatically.

**Comment by matszpk** [comment]
Few days I ran OpenCL 2.0 example (GenericPtr) on Win 8.1 with latest Crimson 16.9.1 beta drivers. It worked correctly (I was testing older version of CLRX). This is problem only on AMDGPU-PRO driver.

**Comment by matszpk** [comment]
Answer at https://community.amd.com/thread/205476. AMDGPU-PRO doesn't support  OpenCL 2.0. This same info is in Releases Notes:
....
 Highlights

        Provides support for Radeon™ RX 480, Radeon™ RX 470 and Radeon™ RX 460 series products
        Supported APIs:
            OpenGL 4.5 and GLX 1.4
            OpenCL 1.2
            Vulkan 1.0
            VDPAU
            Vulkan support for DOTA2
....
CLRX already support OpenCL 2.0 format for AMDGPU-PRO, however driver does not.
I removed my old gfx card from motherboard and I ran ubuntu with AMDGPU-PRO. clinfo returned that my Bonaire supports only OpenCL 1.2.

**Comment by wolf9466** [comment]
So how do I get this to disassemble/assemble as OpenCL 1.2?

**Comment by matszpk** [comment]
I ran VectorAdd example for OpenCL 1.2 and it works good on AMDGPU-PRO. Just, you can try.

**Comment by wolf9466** [comment]
I just forced -cl-std=CL1.2 - and it disassembles as .amdcl2.

**Comment by matszpk** [comment]
oh....  I forgot about one. New OpenCL drivers just use OpenCL 2.0 format while using OpenCL 1.2. This is normal behaviour (same as later Windows drivers). Thus, the CLRX recognizes stored binaries as .amdcl2. Maybe  later, I will find more time to solve this problem.

**Comment by wolf9466** [comment]
I'd be willing to donate, if that would help. I *really* want the ability to use this for my Polaris GPUs on linux.

**Comment by matszpk** [comment]
I apologize for my postponed reaction, but I have *really* urgent task to do, I couldn't properly understand clue of the problem.

**Comment by wolf9466** [comment]
That's okay - some other time.

**Comment by matszpk** [comment]
I fixed detection AMDGPU-PRO driver while compilation (by adding extra path to AMDOpenCL library). Now, samples (with OpenCL 2.0 format, even with option "-cl-std=CL2.0") works correctly. With latest version CLRX detect an OpenCL libraries in 2117.07 version. However, disassembled codes should be work correctly, because disassembler includes info about driver version '.driver_version 203603'.
I successfully disassembled simple code compiled for devices Stoney, Ellesmere, Baffin (Polaris). If last fix doesn't solve your problem, please report me about that.

**Comment by wolf9466** [comment]
Disassembling a binary made with AMD's OCL runtime (AMDGPU-PRO), and then assembling it produces a binary that, if you attempt to load it with AMD's OCL runtime, it throws this error:

[06:19:02] Error -11: Building Program (clBuildProgram)
[06:19:02] Error: AMD HSA Code Object loading failed.

**Comment by matszpk** [comment]
Some hint for you: to disassemble code for assemble you should use options -a (--all) and optionally '-C' (-config). like:
```clrxdisasm -aC binary```

**Comment by wolf9466** [comment]
Just tried -aC - did not work.

**Comment by matszpk** [comment]
Do you check also only '-a' option for clrxdisasm (option '-C' stores kernel configuration in human-readable form, without this option disassembler stores closely this same config as in input file).
I was comparing differences between original binary generated by AMDGPU-PRO driver and reconstructed binary by CLRX disassembler and assembler (for device Ellesmere). I couldn't find important differences which would be have impact while loading file (kernel config and other metadata).
Can you give a binary for simple kernel (including disassembly code and clrxasm binary) (like VectorAdd) to track differences? What a version of CLRX do you use?

**Comment by wolf9466** [comment]
I just downloaded it from git last night. I'll try only -a.

**Comment by wolf9466** [comment]
This is the binary produced by AMD: https://ottrbutt.com/tmp/ethashEllesmeregw256l8.bin
This is the one produced by CLRX: https://ottrbutt.com/tmp/ethashEllesmeregw256l8.bin.clrx
This is the disassembly: https://ottrbutt.com/tmp/ethashEllesmeregw256l8.asm

**Comment by matszpk** [comment]
Thank you. Very likely CLRX assembler stores incorrectly info about constant data (sections, offset, symbols). If I will find free time I will fix that bug.

**Comment by wolf9466** [comment]
Thank you!

**Comment by matszpk** [comment]
Ok. Can you do yet another thing? Just write a very simple kernel (like vectorAdd), try to rebuild with CLRX and load via OpenCL runtime. Just check, whether CLRX can correctly rebuild simpliest kernel for the AMDGPU-PRO driver.

**Comment by matszpk** [comment]
I have some doubts, whether your problems caused by a wrong storing of constant datas by CLRX while building binaries. Can you make simple test described above?

**Comment by matszpk** [comment]
I have no a RX 480, however I was trying load by runtime (AMDGPUPRO) a delivered by you binaries. I modified device (to my Bonaire) type in ELF header (no other changes) on both binaries (CLRX and original). Results are:

* originally generated file by driver couldn't be loaded by my driver installation
* file regenerated by CLRX was succesfully loaded by my driver installation.

I created simple kernel that have this same amount of constant data (192 bytes) and I was trying to load both original binary and regenerated binary (both compiled for Bonaire). Both was loaded successfully by my driver installation.
Maybe AMDGPU-PRO treats in another way binaries for GCN 1.1 and GCN 1.2? I don't know.
